engine/verification/assigner/engine.go

```diff
@@ -147,14 +149,36 @@ func (e *Engine) processChunk(chunk *flow.Chunk, resultID flow.Identifier) {
 		return
 	}
 
+	e.metrics.OnChunkProcessed()
+
 	// notifies chunk queue consumer of a new chunk
 	e.newChunkListener.Check()
 	log.Debug().Msg("chunk locator successfully pushed to chunks queue")
 }
 
-// ProcessFinalizedBlock indexes the execution receipts included in the block, and handles their chunk assignments.
-// Once it is done handling all the receipts in the block, it notifies the block consumer.
+// ProcessFinalizedBlock is the entry point of assigner engine. It pushes the block down the pipeline with tracing on it enabled.
+// Through the pipeline the execution receipts included in the block are indexed, and their chunk assignments are done, and
+// the assigned chunks are pushed to the chunks queue, which is the output stream of this engine.
+// Once the assigner engine is done handling all the receipts in the block, it notifies the block consumer.
 func (e *Engine) ProcessFinalizedBlock(block *flow.Block) {
+	blockID := block.ID()
+	span, ok := e.tracer.GetSpan(blockID, trace.VERProcessFinalizedBlock)
+	ctx := context.Background()
+	if !ok {
+		span = e.tracer.StartSpan(blockID, trace.VERProcessFinalizedBlock)
+		span.SetTag("block_id", blockID)
+		defer span.Finish()
+	}
+	ctx = opentracing.ContextWithSpan(ctx, span)
+
+	e.tracer.WithSpanFromContext(ctx, trace.VERAssignerHandleFinalizedBlock, func() {
+		e.processFinalizedBlock(ctx, block)
+	})
```

Comment:
all of this should be behind a check if tracer is not nil to make things faster

Reply:
@ramtinms would you please explain further why this check is necessary? and what should we do if `tracer` is `nil`? I am currently under the impression that `nil` tracer means a misconfiguration in our bootstrapping, so shouldn't we let the code panic (immediately at line 165) if `tracer` is `nil`? 

And if the `nil` checking is necessary, are we going to do that on every invocation in the subsequent methods of this engine?